src/masternodes/mn_checks.cpp

```diff
@@ -1452,6 +1461,137 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
+            const auto disabled = attributes->GetValue(tokenKey, false);
+            if (disabled) {
+                return Res::Err("DFIP2203 currently disabled for token %d", obj.destination);
+            }
+
+            const auto loanToken = mnview.GetLoanTokenByID({obj.destination});
+            if (!loanToken) {
+                return Res::Err("Could not get destination loan token %d. Set valid destination.", obj.destination);
+            }
+        } else {
+            if (obj.destination != 0) {
+                return Res::Err("Destination should not be set when source amount is a dToken");
+            }
+
+            CDataStructureV0 tokenKey{AttributeTypes::Token, obj.source.nTokenId.v, TokenKeys::DFIP2203Disabled};
+            const auto disabled = attributes->GetValue(tokenKey, false);
+            if (disabled) {
+                return Res::Err("DFIP2203 currently disabled for token %s", obj.source.nTokenId.ToString());
+            }
+        }
+
+        const auto resVal = GetFutureSwapContractAddress();
+        if (!resVal) {
+            return resVal;
+        }
+
+        CDataStructureV0 liveKey{AttributeTypes::Live, ParamIDs::Economy, EconomyKeys::DFIP2203Tokens};
+        auto balances = attributes->GetValue(liveKey, CBalances{});
+
+        if (obj.withdraw) {
+            const auto blockPeriod = attributes->GetValue(blockKey, CAmount{});
+            const uint32_t startHeight = height - (height % blockPeriod);
+            std::map<CFuturesUserKey, CFuturesUserValue> userFuturesValues;
+
+            mnview.ForEachFuturesUserValues([&](const CFuturesUserKey& key, const CFuturesUserValue& futuresValues) {
+                if (key.height <= startHeight) {
+                    return false;
+                }
+
+                if (key.owner == obj.owner &&
+                    futuresValues.source.nTokenId == obj.source.nTokenId &&
+                    futuresValues.destination == obj.destination) {
+                    userFuturesValues[key] = futuresValues;
+                }
+
+                return true;
+            }, {height, obj.owner, std::numeric_limits<uint32_t>::max()});
+
+            CTokenAmount totalFutures{};
+            totalFutures.nTokenId = obj.source.nTokenId;
+
+            for (const auto& [key, value] : userFuturesValues) {
+                totalFutures.Add(value.source.nValue);
+                mnview.EraseFuturesUserValues(key);
+            }
+
+            auto res = totalFutures.Sub(obj.source.nValue);
+            if (!res) {
+                return res;
+            }
+
+            if (totalFutures.nValue > 0) {
+                auto res = mnview.StoreFuturesUserValues({height, obj.owner, txn}, {totalFutures, obj.destination});
+                if (!res) {
+                    return res;
+                }
+            }
+
+            res = TransferTokenBalance(obj.source.nTokenId, obj.source.nValue, *resVal, obj.owner);
```

Comment:
missing check that obj.source.nValue <= totalFutures.nValue.
Otherwise it would withdraw more than the user send to the contract, right?

Reply:
That is checked here.

https://github.com/DeFiCh/ain/pull/1155/files#diff-5b5f5b126aad5172f5e15066f2926f4afb57e75eb624b509f56233c14cd4b5c9R1553

Reply:
Ah right, there is a check within Sub. thx.